server/src/main/java/org/opensearch/index/mapper/DefaultDerivedFieldResolver.java

```diff
@@ -189,9 +190,10 @@ private void initDerivedFieldTypes(Map<String, Object> derivedFieldsObject, List
 
     private Map<String, DerivedFieldType> getAllDerivedFieldTypeFromObject(Map<String, Object> derivedFieldObject) {
         Map<String, DerivedFieldType> derivedFieldTypes = new HashMap<>();
+        // deep copy of derivedFieldObject is required as DocumentMapperParser modifies the map
         DocumentMapper documentMapper = queryShardContext.getMapperService()
             .documentMapperParser()
-            .parse(DerivedFieldMapper.CONTENT_TYPE, derivedFieldObject);
+            .parse(DerivedFieldMapper.CONTENT_TYPE, IngestDocument.deepCopyMap(derivedFieldObject));
```

Comment:
`IngestDocument` is pretty specifically related to ingest pipelines, I think. Probably best to refactor that `deepCopyMap` into some sort of properly common utility versus using `IngestDocument` here.

Reply:
The closest class I could find is Maps.java - https://github.com/opensearch-project/OpenSearch/blob/ad7ce4cd446523d52286530ccfbb1dabc2dc7f86/server/src/main/java/org/opensearch/common/util/Maps.java#L43
I want to move the deepCopy(Object) method - https://github.com/opensearch-project/OpenSearch/blob/8e32ed736372aa90db4c0ce3b85888b7b473a337/server/src/main/java/org/opensearch/ingest/IngestDocument.java#L760  - which is generic for an object and not just Map, so Maps.java doesn't look like the right place for it.

Reply:
Do you actually need a true recursive deep copy here, versus just doing `new HashMap(derivedFieldObject)`?

As to your specific question, looking more closely I'm hesitant to move that `deepCopy(Object)` method to a generic utility. Deep copying an arbitrary Java object isn't really feasible, and that method is actually quite limited on what it will do. It can only deep copy scalar values, Lists or Maps of scalar values (but not Sets or Collections), byte arrays, and Date and ZonedDateTime instances. Basically, I would not recommend using it outside the context of an IngestDocument.

Reply:
@andrross  yes, a deep copy is needed here as `derivedFieldObject` map is created by parsing a json, so all types supported in json objects are valid here. The `deepCopy()` method here, will take care of all the cases for derived fields and in general json. Also, this method seems safe as it throws IllegalArgumentException is the type isn't support instead of trying to shallow copy them. 
What is your suggestion here?

Reply:
Hmm, the problem is that the `org.opensearch.common.util.CopyUtils::deepCopy(Object)` signature looks incredibly generic, but is in fact quite limited and is pretty specific about what it will do. `IngestDocument::deepCopy` was okay because it is clearly scoped to an IngestDocument. But now you're using it on the search path. So what is the concept in common here between this search path and the IngestDocument? I don't think it is strictly "JSON types" because OpenSearch generally deals with XContent and not JSON, and I don't think there is a true canonical mapping between JSON and Java types. Also at least one code path looks like the source of this map is [the transport protocol](https://github.com/opensearch-project/OpenSearch/blob/9da6170fc38653284cfeee0e5c1800d41999ba08/server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java#L294). Also am I reading that code correctly that it may return a [set or collection type](https://github.com/opensearch-project/OpenSearch/blob/2a37bddefe876a91ace2faf1a3cc91ebe0b7b6e5/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java#L786-L789)?

Reply:
@andrross thanks for taking a deeper on deepCopy :)

> So what is the concept in common here between this search path and the IngestDocument?

The only reason to use it was because it generic enough for `derivedFieldsObject` cases a and I agree that it is not a generic deep copy Java object, which isn't possible anyways. 

> I don't think it is strictly "JSON types" because OpenSearch generally deals with XContent and not JSON, and I don't think there is a true canonical mapping between JSON and Java types

It is created by parsing XContent and not JSON. There isn't any canonical mapping between two of them but we always try to map XContent to Java types depending on the underlying use case -  the mapping logic is different for a document parser than from a field type parser. 

For derived field use case, the `deepCopy()` method used here is holds valid for any current or future use cases. Translation of XContent to final mapper happens in 2 parts, first converting it to a regular java Map and then passing the individual key values to DerivedFieldMapper to create the field type. 

I can think of 2 options - 1) create a duplicate of this method in `DefaultDerivedFieldResolver` with a private access 2) use `CopyUtils` with a better documentation to avoid misuse. If you have better suggestion, i'm willing to evaluate it.


> Also at least one code path looks like the source of this map is [the transport protocol](https://github.com/opensearch-project/OpenSearch/blob/9da6170fc38653284cfeee0e5c1800d41999ba08/server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java#L294). Also am I reading that code correctly that it may return a [set or collection type](https://github.com/opensearch-project/OpenSearch/blob/2a37bddefe876a91ace2faf1a3cc91ebe0b7b6e5/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java#L786-L789)?

It cannot decoded a set for anything encoded as a map? It should probably use https://github.com/opensearch-project/OpenSearch/blob/2a37bddefe876a91ace2faf1a3cc91ebe0b7b6e5/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java#L761 ?

Reply:
Ideally, I would have liked this to have an option to avoid mutating the `node` map passed to it - https://github.com/opensearch-project/OpenSearch/blob/8ca3e6ad51d73b6b89cb88eca8f4c759f708e59b/server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java#L262
But this seems like a bigger change as node and subNodes are passed down to individual mappers too which can modify them.

Reply:
> It cannot decoded as a set for anything encoded as a map before?

Do you know the structure of this map you're copying? How do you know that somewhere in the structure of `Map<String, Object>` that there isn't a set or a list type?

Reply:
> Ideally, I would have liked this to have an option to avoid mutating the node map passed to it

It does seem like we're working around code that was fundamentally meant for non-concurrent use. I don't have enough depth in this particular area to know the scope of avoiding mutating the parameters.

Reply:
Sorry to rathole on this seemingly small issue :)

I'd be happy with either of the following:

- Rename CopyUtils to something more specific. Probably move it out of `lib/common` into either `lib/core` or just keep it in server. The risk is that future devs try to use this as a general purpose deep copy, which could lead to subtle bugs.
- Reimplement deepCopy as a private method similar to below (provided the simpler version would work here):
```
private static Object deepCopyMaps(Object value) {
    if (value instanceof Map) {
        Map<?, ?> mapValue = (Map<?, ?>) value;
        Map<Object, Object> copy = new HashMap<>(mapValue.size());
        for (Map.Entry<?, ?> entry : mapValue.entrySet()) {
            copy.put(entry.getKey(), deepCopyMaps(entry.getValue()));
        }
        return copy;
    }
    return value;
}
```

Reply:
> Do you know the structure of this map you're copying? How do you know that somewhere in the structure of Map<String, Object> that there isn't a set or a list type?

List will be present but will be covered by these https://github.com/opensearch-project/OpenSearch/blob/2a37bddefe876a91ace2faf1a3cc91ebe0b7b6e5/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java#L755-L757

XContent parser will return one of the following value type after parsing it as a map and they all seem to be covered with this logic of deepCopy(). 
https://github.com/opensearch-project/OpenSearch/blob/e2f21d824ac0dc76d826b9bbdc32ec80df7c9bf6/libs/core/src/main/java/org/opensearch/core/xcontent/AbstractXContentParser.java#L406

and this is how its identifying the the token type - https://github.com/opensearch-project/OpenSearch/blob/301c766075a4a91870439482cab4cfb09d9299bd/libs/x-content/src/main/java/org/opensearch/common/xcontent/json/JsonXContentParser.java#L243

>> 
```java
private static Object deepCopyMaps(Object value) {
    if (value instanceof Map) {
        Map<?, ?> mapValue = (Map<?, ?>) value;
        Map<Object, Object> copy = new HashMap<>(mapValue.size());
        for (Map.Entry<?, ?> entry : mapValue.entrySet()) {
            copy.put(entry.getKey(), deepCopyMaps(entry.getValue()));
        }
        return copy;
    }
    return value;
}
```
This might not work as it can contain a list/array type. But let me simplify it a bit and keep it private.

Reply:
I have tried to simplify it but still kept the check for `value instanceof byte[]` as I'm unsure of its use for derived field use case.